File: s16/go16/s16.go
package main

type CPU struct {
	reg       *Registers
	ram       *Memory
	instCount int
}

const (
	A, _ = iota, iota
	B, _
	PC, ADD
	PCZ, AND
	PCN, OR
	PCC, XOR
	PCZC, RSH
	PCZD, LIT
	CP, _
	C, _
	PlusC, Cminus
	MinusC, Cplus
	DP, _
	D, _
	Dplus, MinusD
	Dminus, PlusD
	CARRY, _
)

func (c *CPU) Decode(instruction Word) {
	c.copy(Addr((instruction>>12)&0x0f), Addr((instruction>>8)&0x0f))
	c.copy(Addr((instruction>>4)&0x0f), Addr(instruction&0x0f))
}

func (c *CPU) copy(dst Addr, src Addr) {
	c.instCount++
	
	val := c.reg.read[src]()
	
	c.reg.write[dst](val)
}

func (c *CPU) PC() Word {
	return c.reg.value[PC]
}

func (c *CPU) GOTO(address Word) {
	c.reg.value[PC] = address
}

func (c *CPU) Exec() {
	pc := c.PC()
	instruction := c.ram.get(pc)
	c.GOTO(pc + 1)
	c.Decode(instruction)
}

func NewCPU() *CPU {
	var reg = NewRegisters()
	var RAM = NewMemory()

	reg.value[CARRY] = 0
	reg.R(A, func() Word {
		return reg.value[A]
	})
	reg.R(B,func() Word {
		return reg.value[B]
	})
	reg.W(A, func(v Word) {
		var calc int32

		reg.value[A] = v;
		
		calc = int32(reg.value[A]) + int32(reg.value[B])
		if calc > 0xffff {
			reg.value[CARRY] = 1
		} else {
			reg.value[CARRY] = 0
		}
	})
	reg.W(B, func(v Word) {
		var calc int32

		reg.value[B] = v;
		
		calc = int32(reg.value[A]) + int32(reg.value[B])
		if calc > 0xffff {
			reg.value[CARRY] = 1
		} else {
			reg.value[CARRY] = 0
		}
	})
	reg.R(ADD, func() Word {
		return reg.value[A]+reg.value[B]
	})
	reg.R(AND, func() Word { return reg.value[A]&reg.value[B] })
	reg.R(OR, func() Word { return reg.value[A]|reg.value[B] })
	reg.R(XOR, func() Word { return reg.value[A]^reg.value[B] })
	reg.R(RSH, func() Word { return reg.value[A]>>1 })
	
	reg.W(PC, func(v Word) {
		reg.value[PC] = v
	})
	reg.W(PCZ, func(v Word) {
		if (reg.value[A] + reg.value[B]) == 0 {
			reg.value[PC] = v
		}
	})
	reg.W(PCN, func(v Word) {
		if (reg.value[A] + reg.value[B]) != 0 {
			reg.value[PC] = v
		}
	})
	reg.W(PCC, func(v Word) {
		if reg.value[CARRY] != 0 {
			reg.value[PC] = v
		}
	})
	reg.W(PCZC, func(v Word) {
		if RAM.get(reg.value[CP]) != 0 {
			reg.value[PC] = v
		}
	})
	reg.W(PCZD, func(v Word) {
		if RAM.get(reg.value[DP]) != 0 {
			reg.value[PC] = v
		}
	})
	
	reg.Plain(CP);
	reg.R(C, func() Word { return RAM.get(reg.value[CP]) })
	reg.W(C, func(v Word) { RAM.set(reg.value[CP], v) })
	reg.R(Cminus, func() Word {
		reg.value[CP]--
		return RAM.get(reg.value[CP] + 1)
	})
	reg.W(PlusC, func(v Word) {
		reg.value[CP]++
		RAM.set(reg.value[CP], v)
	})
	reg.R(Cplus, func() Word {
		reg.value[CP]++
		return RAM.get(reg.value[CP] - 1)
	})
	reg.W(MinusC, func(v Word) {
		reg.value[CP]--
		RAM.set(reg.value[CP], v)
	})
	reg.R(LIT, func() Word {
		reg.value[PC]++
		return RAM.get(reg.value[PC] - 1)
	})

	reg.Plain(DP);
	reg.R(D, func() Word { return RAM.get(reg.value[DP]) })
	reg.W(D, func(v Word) { RAM.set(reg.value[DP], v) })
	reg.R(MinusD, func() Word {
		reg.value[DP]--
		return RAM.get(reg.value[DP])
	})
	reg.W(Dplus, func(v Word) {
		reg.value[DP]++
		RAM.set(reg.value[DP] - 1, v)
	})
	reg.R(PlusD, func() Word {
		reg.value[DP]++
		return RAM.get(reg.value[DP])
	})
	reg.W(Dminus, func(v Word) {
		reg.value[DP]--
		RAM.set(reg.value[DP] + 1, v)
	})
	
	reg.R(LIT, func() Word {
		literal := RAM.get(reg.value[PC])
		reg.value[PC]++
		return literal
	})


	return &CPU{
		reg: reg,
		ram: RAM,
	}
}

File: s16/c16/s16.c
#include "registers.h"
#include "memory.h"
#include "datatypes.h"
#include <stdio.h>

#define A    0x00
#define B    0x01
#define PC   0x02
#define ADD  0x02
#define PCZ  0x03
#define AND  0x03
#define PCN  0x04
#define OR   0x04
#define PCC  0x05
#define XOR  0x05
#define PCZC 0x06
#define RSH  0x06
#define PCZD 0x07
#define LIT  0x07
#define CP    0x08
#define C     0x09
#define INC_C 0x0a
#define C_DEC 0x0a
#define DEC_C 0x0b
#define C_INC 0x0b
#define DP    0x0c
#define D     0x0d
#define D_INC 0x0e
#define DEC_D 0x0e
#define D_DEC 0x0f
#define INC_D 0x0f

int instCount;

int carry () {
	int32_t sum;
	
	sum = Val[A] + Val[B];
	
	if (sum > 0xffff) {
		return 1;
	}
	return 0;
}

void writeA (word val) {
	Val[A] = val;
}
word readA () {
	return Val[A];
}

void writeB (word val) {
	Val[B] = val;
}
word readB () {
	return Val[B];
}

word add () {
	return Val[A] + Val[B];
}
word and () {
	return Val[A] & Val[B];
}
word or () {
	return Val[A] | Val[B];
}
word xor () {
	return Val[A] ^ Val[B];
}
word rsh () {
	return Val[A] >> 1;
}
word lit () {
	word ret;
	ret = RAM[Val[PC]];
	Val[PC]++;
	return ret;
}

void pc (word val) {
	Val[PC] = val;
}
void pcz (word val) {
	if (!(Val[A]+Val[B])) {
		Val[PC] = val;
	}
}
void pcn (word val) {
	if (Val[A]+Val[B]) {
		Val[PC] = val;
	}
}
void pcc (word val) {
	if (carry()) {
		Val[PC] = val;
	}
}
void pczc (word val) {
	if (!(Val[C])) {
		Val[PC] = val;
	}
}
void pczd (word val) {
	if (!(Val[D])) {
		Val[PC] = val;
	}
}

word readCP () {
	return Val[CP];
}
void writeCP (word val) {
	Val[CP] = val;
}
word readC () {
	return RAM[Val[CP]];
}
void writeC (word val) {
	setRAM(Val[CP], val);
}
word c_dec () {
	word ret;
	ret = RAM[Val[CP]];
	Val[CP]--;
	return ret;
}
void inc_c (word val) {
	Val[CP]++;
	setRAM(Val[CP], val);
}
word c_inc () {
	word ret;
	ret = RAM[Val[CP]];
	Val[CP]++;
	return ret;
}
void dec_c (word val) {
	Val[CP]--;
	setRAM(Val[CP], val);
}

word readDP () {
	return Val[DP];
}
void writeDP (word val) {
	Val[DP] = val;
}
word readD () {
	return RAM[Val[DP]];
}
void writeD (word val) {
	setRAM(Val[DP], val);
}
word dec_d () {
	Val[DP]--;
	return RAM[Val[DP]];
}
void d_inc (word val) {
	setRAM(Val[DP], val);
	Val[DP]++;
}
word inc_d () {
	Val[DP]++;
	return RAM[Val[DP]];
}
void d_dec (word val) {
	setRAM(Val[DP], val);
	Val[DP]--;
}

void initCPU () {
	initRAM();
	initRegisters();
	instCount = 0;
	
	R[A] = readA;
	W[A] = writeA;
	R[B] = readB;
	W[B] = writeB;
	R[ADD] = add;
	W[PC] = pc;
	R[AND] = and;
	W[PCZ] = pcz;
	R[OR] = or;
	W[PCN] = pcn;
	R[XOR] = xor;
	W[PCC] = pcc;
	R[RSH] = rsh;
	W[PCZC] = pczc;
	R[LIT] = lit;
	W[PCZD] = pczd;
	R[CP] = readCP;
	W[CP] = writeCP;
	R[C] = readC;
	W[C] = writeC;
	R[C_DEC] = c_dec;
	W[INC_C] = inc_c;
	R[C_INC] = c_inc;
	W[DEC_C] = dec_c;
	R[DP] = readDP;
	W[DP] = writeDP;
	R[D] = readD;
	W[D] = writeD;
	R[DEC_D] = dec_d;
	W[D_INC] = d_inc;
	R[INC_D] = inc_d;
	W[D_DEC] = d_dec;
}

void copy (address dst, address src) {
	instCount++;
	(*W[dst])((*R[src])());
}

void decode (word instruction) {
	copy((instruction>>12) & 0xf, (instruction>>8) & 0xf);
	copy((instruction>>4) & 0xf, instruction & 0xf);
}

void exec() {
	word instruction;
	
	instruction = RAM[Val[PC]];
	Val[PC]++;
	
	decode(instruction);
}
